### PR TITLE
Hillary Stewart

### PR DESCRIPTION
**Pull request by hillstew**

### CODE REVIEW

File: test/joke_test.rb [hunk 1]
```diff
@@ -3,9 +3,16 @@
...
+  end
+
   def test_it_exists
+    assert_instance_of Joke, @joke
   end
 
   def test_it_has_attributes
+    assert_equal "Why did the strawberry cross the road?", @joke.setup
+    assert_equal "Because his mother was in a jam.", @joke.punchline
   end
```
Comment (iandouglas) [inline test/joke_test.rb:17]: good job in here, only thing you missed was testing that the joke had an ID value

File: lib/joke.rb [hunk 2]
```diff
@@ -1,2 +1,9 @@
 class Joke
+  attr_reader :id, :setup, :punchline
+
+  def initialize(id, setup, punchline)
+    @id = id
+    @setup = setup
+    @punchline = punchline
+  end
```
Comment (iandouglas) [inline lib/joke.rb:8]: good job in here

File: test/user_test.rb [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+require './lib/user'
+require './lib/joke'
+
+class UserTest < Minitest::Test
+  def setup
+    @sal = User.new("Sal")
+    @ali = User.new("Ali")
+    @joke1 = Joke.new(1, "Why did the strawberry cross the road?", "Because his mother was in a jam.")
+    @joke2 = Joke.new(2, "How do you keep a lion from charging?", "Take away its credit cards.")
+  end
```
Comment (iandouglas) [inline test/user_test.rb:12]: great setup

File: test/user_test.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+  def test_it_exists
+    assert_instance_of User, @sal
+  end
+
+  def test_it_has_attributes
+    assert_equal "Sal", @sal.name
+  end
```
Comment (iandouglas) [inline test/user_test.rb:20]: good job here, be sure to test all initialized attributes, like asserting that Sal doesn't know any jokes yet

File: test/user_test.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+    @sal.tell(@ali, @joke1)
+    @sal.tell(@ali, @joke2)
+    expected = @ali.joke_by_id(2)
+    assert_equal expected, @joke2
+  end
+
+  def test_it_can_learn_a_joke
+    @ali.learn(@joke1)
+    @ali.learn(@joke2)
+    assert_equal [@joke1, @joke2], @ali.jokes
```
Comment (iandouglas) [inline test/user_test.rb:38]: good job here

File: test/user_test.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def test_it_has_attributes
+    assert_equal "Sal", @sal.name
+  end
+
+  def test_it_can_tell_a_joke
+    @sal.tell(@ali, @joke1)
+    @sal.tell(@ali, @joke2)
+    assert_equal [@joke1, @joke2], @ali.jokes
```
Comment (iandouglas) [inline test/user_test.rb:25]: good start in here ... how can Sal tell Ali a joke that he hasn't learned yet, though?

File: test/user_test.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def test_it_can_return_joke_by_id
+    @sal.tell(@ali, @joke1)
+    @sal.tell(@ali, @joke2)
+    expected = @ali.joke_by_id(2)
+    assert_equal expected, @joke2
```
Comment (iandouglas) [inline test/user_test.rb:32]: great job here

File: test/user_test.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+  end
+
+  def test_it_can_learn_a_joke
+    @ali.learn(@joke1)
+    @ali.learn(@joke2)
+    assert_equal [@joke1, @joke2], @ali.jokes
+end
```
Comment (iandouglas) [inline test/user_test.rb:39]: Overall, very good testing in here. Be sure to add a blank line before your assertions for easier readability.

File: lib/user.rb [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+  attr_reader :name, :jokes
+
+  def initialize(name)
+    @name = name
+    @jokes = []
+  end
+
+  def tell(user, joke)
+    user.jokes << joke
+    @jokes
```
Comment (iandouglas) [inline lib/user.rb:11]: Generally we don't want one object to directly inject data into another object. Is there another method we could reuse here instead?
Also, the last line, `@jokes` wasn't necessary, as our interaction pattern didn't explicitly ask that you return anything specific. It's okay that you added it.

File: lib/user.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  end
+
+  def joke_by_id(id)
+    @jokes.find do |joke|
+      joke.id == id
+    end
```
Comment (iandouglas) [inline lib/user.rb:17]: great job here

File: test/open_mic_test.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    assert_instance_of OpenMic, @open_mic
+  end
+
+  def test_it_has_attributes
+    assert_equal "Comedy Works", @open_mic.location
+  end
+
+  def test_it_has_performers
+    assert_equal [], @open_mic.performers
+  end
```
Comment (iandouglas) [inline test/open_mic_test.rb:26]: great job up to here. You can combine these last two tests into a `test_has_attributes` method to check anything you initialized

File: test/open_mic_test.rb [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+  def test_it_can_welcome_performers
+    @open_mic.welcome(@sal)
+    @open_mic.welcome(@ali)
+    assert_equal [@sal, @ali], @open_mic.performers
```
Comment (iandouglas) [inline test/open_mic_test.rb:31]: good job here

File: test/open_mic_test.rb [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+    @open_mic.welcome(@ali)
+    @ali.learn(@joke1)
+    @ali.learn(@joke2)
+
+    assert_equal false, @open_mic.repeated_jokes?
+
+    @ali.tell(@sal, @joke1)
+    
+    assert_equal true, @open_mic.repeated_jokes?
+  end
```
Comment (iandouglas) [inline test/open_mic_test.rb:45]: great job here

File: lib/open_mic.rb [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    @performers.each do |performer|
+      performer.jokes.each do |joke|
+        if !jokes.include?(joke)
+          jokes << joke
+        elsif jokes.include?(joke)
+          return true
+        end
+      end
+    end
+    false
```
Comment (iandouglas) [inline lib/open_mic.rb:25]: good work here, we'll talk about a way to shorten this in the walkthrough; I really like how you shortcut processing any further with an explicit `return` on line 21

**Review by iandouglas — COMMENTED** [review]
great job

**Comment by hillstew** [comment]
Thank you for the feedback! Look forward to the walkthrough tomorrow 😊

**Pull request closed without merging**